doc/explanation/mental-model.rst

```diff
@@ -0,0 +1,175 @@
+
+.. code:: dune
+
+  (rule
+   (deps convert/json2ml.exe config.json)
+   (target config.ml)
+   (action
+    (run convert/json2ml.exe config.json -o config.ml)))
+
+This rule will:
+
+- read its dependencies: ``convert/json2ml.exe`` and ``config.json``
+- and write its target: ``config.ml``
+- using an action: ``(run convert/json2ml.exe config.json -o config.ml)``
+
+This rule is very explicit: we write a stanza for a single Dune rule.
+
+Building the Executable
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In contrast, to describe the compilation of the executable, we would write:
+
+.. code:: dune
+
+  (executable
+   (name tool)
+   (modules main config))
+
+Here, we use Dune's abstractions. It knows about the OCaml compilation model:
+the modules need to be compiled and linked together. So it will generate the
+following rules under the hood:
+
+- one rule to compile the ``Main`` module:
+
+  - it will read its dependency: ``main.ml``
+  - and write its output: ``main.cmx``
+  - using an action: ``(run ocamlopt -c main.ml)``
+
+- one rule to compile the ``Config`` module:
+
+  - it will read its dependency: ``config.ml``
+  - and write its output: ``config.cmx``
+  - using an action: ``(run ocamlopt -c config.ml)``
+
+- one rule to compile the ``tool.exe`` executable:
+
+  - it will read its dependencies: ``main.cmx`` and ``config.cmx``
+  - and write its output: ``tool.exe``
+  - using an action: ``(run ocamlopt -o tool.exe main.cmx config.cmx``)
+
+Note that in this example, some files are targets of a rule and dependencies of
+another (``.cmx`` files). We are unlikely to ever interact with them directly,
+so it can also be useful to think of the ``(executable)`` stanza as acting as a
+big rule with ``main.ml`` and ``config.ml`` as inputs.
+
+Running the Tests
+-----------------
+
+Some rules do not produce any output file, but we're still interested in
+running an action. A test is a good example: we want the build process to exit
+with an error code if the action fails.
+In that case, the rule does not have targets, but we "attach" it to an
+:term:`alias`, ``runtest`` in this case.
+
+.. code:: dune
+
+  (rule
+   (deps tool.exe testdata.txt)
+   (alias runtest)
+   (action
+    (run tool.exe testdata.txt)))
+
+This rule:
+
+- reads its dependencies: ``tool.exe`` and ``testdata.txt``
+- writes no targets
+- by running its action: ``(run tool.exe testdata.txt)``
+- (but it is attached to ``runtest``)
+
+How Dune Interprets Rules
+-------------------------
+
+Now that we have seen how Dune sets up rules for a project, let's see how it
+executes a ``dune build`` command.
+
+Dune can build targets: either files, or aliases. ``dune build tool.exe`` will
+build ``tool.exe``, while ``dune build @runtest`` will build the ``runtest``
+alias (``dune runtest`` is a shortcut for ``dune build @runtest``).
+
+To build the target *t*, Dune will first check is *t* is in the source tree. In
+that case, it will copy it to the ``_build`` directory. Otherwise, it will
```

Comment:
I think "it will copy it to the `_build` directory" may be leaking some unnecessary implementation details. Without thinking deeply about how Dune builds are working, copying sources into the `_build` directory might seem unnecessary/confusing. In fact, internally, this will soon not be true in some cases.

Reply:
So, maybe, "In this case, there is nothing to do" instead?